Lib/importlib/abc.py

```diff
@@ -15,20 +15,36 @@
 import abc
 import warnings
 
-# for compatibility with Python 3.10
-from .resources.abc import ResourceReader, Traversable, TraversableResources
+from .resources import abc as _resources_abc
 
 
 __all__ = [
     'Loader', 'Finder', 'MetaPathFinder', 'PathEntryFinder',
     'ResourceLoader', 'InspectLoader', 'ExecutionLoader',
     'FileLoader', 'SourceLoader',
-
-    # for compatibility with Python 3.10
-    'ResourceReader', 'Traversable', 'TraversableResources',
 ]
 
 
+def __getattr__(name, canonical=_resources_abc):
+    """
+    For backwards compatibility, continue to make names
+    from canonical available through this module.
+    """
+    if name in canonical.__all__:
+        obj = getattr(canonical, name)
+        import warnings
+        warnings.warn(
+            f"Using or importing the ABCs from {__name__!r} instead "
+            f"of from {canonical.__name__!r} is now deprecated, "
+            "scheduled for removal in Python 3.13",
+            DeprecationWarning,
+            stacklevel=2,
+        )
+        globals()[name] = obj
+        return obj
+    raise AttributeError(f'module {__name__!r} has no attribute {name!r}')
```

Comment:
I suppose it's technically possible for a name to be in `importlib.resources.abc.__all__` but not be in the module.  That's not the case today and hopefully won't be the case in the future, but it *is* possible.  The converse may also be true (i.e. name in the module, but missing from `__all__` though that wouldn't be a backward compatibility issue).

So, for example, the `getattr(canonical, name)` *could* fail, but then you'd still get an `AttributeError`, just with a different message.  

Questions:

- Is the special messaged `AttributeError` at the end of this function worth it?
- Should you catch any potential `AttributeError`s coming from the `getattr()` in this function?
- Can you just use `resources_abc` directly?
- Instead of comparing `name in canonical.__all__` how about just hardcoding `["ResourceReader", "Traversable", "TraversableResources"]` since those are the *only* names that are subject to backward compatibility.
- Can you add a comment and issue for all of this to be removed in 3.13 so we don't forget?

Reply:
> I suppose it's technically possible for a name to be in `importlib.resources.abc.__all__` but not be in the module. That's not the case today and hopefully won't be the case in the future, but it _is_ possible. The converse may also be true (i.e. name in the module, but missing from `__all__` though that wouldn't be a backward compatibility issue).

I thought about this and don't expect there to be a lot of change in these modules. I adopted this technique from [the one used by collections](https://github.com/python/cpython/commit/c66f9f8d3909f588c251957d499599a1680e2320), and the technique worked there. While it's true that it could potentially expose more than was previously exposed, it would be doing so with a deprecation warning, so would actively discourage use of that name.

I consider a value in `__all__` that's not present in the module to be a bug and such a bug would be caught by linters (for importlib_resources), so I'm not worried about it.

> So, for example, the `getattr(canonical, name)` _could_ fail, but then you'd still get an `AttributeError`, just with a different message.
> 
> Questions:
> 
>     * Is the special messaged `AttributeError` at the end of this function worth it?

I didn't consider the `AttributeError` to have a special message. I don't know how else to deal with an attribute name that doesn't match. We want it to raise an `AttributeError`. Are you suggesting not to supply a message and to just raise a non-descript `AttributeError()`?

>     * Should you catch any potential `AttributeError`s coming from the `getattr()` in this function?

I don't think so. The name was already found in `__all__` and it would be a bug if that attribute didn't exist, so it would be overkill to trap an exception that will never happen.

>     * Can you just use `resources_abc` directly?

I can, but this form is more portable and abstracts away the concept of the canonical module for easier re-use.

>     * Instead of comparing `name in canonical.__all__` how about just hardcoding `["ResourceReader", "Traversable", "TraversableResources"]` since those are the _only_ names that are subject to backward compatibility.

I was aiming to generalize (just slightly) the technique used for `collections.abc` deprecations. I was even tempted to go a step further and provide a fully-generalized implementation that could be used for future attribute deprecations, but decided not to do that because I'm not sure if it's useful and it would be a burden to make that implementation available for both importlib.resources and its backport.

I'm really pleased that the body of this `__getattr__` makes no references to anything specific about what's being deprecated, but instead reflects the essential canonical names.

>     * Can you add a comment and issue for all of this to be removed in 3.13 so we don't forget?

In my opinion, the code itself serves as an indication that the deprecated behavior needs to be removed. But yes, I do see the value in somehow indicating in the code the versions where this landed and is expected to change. I'll want to add that to `importlib_resources` first and then sync it here. As I think about it more, adding a "Python version" where this functionality was deprecated could be problemmatic, because the backport works across Python versions and at a different cadence. Still, I agree we need something to track the follow-up work.

As for the issue, I'm slightly reluctant to file an issue that's unactionable for over a year, unless there's a process to help manage those delayed issues. For example, there's not even a 3.13 tag for issues yet (and no other tags I see that are relevant for "deferred" except for blockers). If there was a "deferred" tag and a "3.13" tag, I'd suggest to create an issue with those tags, close it, and then when 3.13 is open for development, re-open those issues.

It occurs to me that such a change could be introduced today if only Python had a means to submit pull requests for future Python versions. I filed https://github.com/python/core-workflow/issues/468 to track that idea.

Reply:
> I'll want to add that to `importlib_resources` first and then sync it here.

I realize now this deprecation isn't necessary for the backport because the backport never exposed these names except in the `abc` module. This change only affects code _only_ in CPython.

Reply:
In 85259269b0, (based on Hugo's suggestion), the code now uses the `warnings._deprecated` function to issue a more general deprecation warning, but also to raise a RuntimeError when Python 3.14b hits, so will serve as a reminder to re-open the now linked bug #93963, which we can re-use for the removal.

Reply:
In 529e6ea178, I remove the generalization and simply repeat `_resources_abc` throughout.

Reply:
#94528 implements the removal (when appropriate).

Reply:
Thanks for the explanations and for the update to this function.  It all seems good to me now.